syntax = "proto3";

package ray.streaming.queue.protobuf;

enum StreamingQueueMessageType {
  StreamingQueueDataMsgType = 0;
  StreamingQueueCheckMsgType = 1;
  StreamingQueueCheckRspMsgType = 2;
  StreamingQueueNotificationMsgType = 3;
  StreamingQueueTestInitMsgType = 4;
  StreamingQueueTestCheckStatusRspMsgType = 5;
  StreamingQueuePullRequestMsgType = 6;
  StreamingQueuePullResponseMsgType = 7;
  StreamingQueueResendDataMsgType = 8;
}

enum StreamingQueueError {
  OK = 0;
  QUEUE_NOT_EXIST = 1;
  DATA_LOST = 2;
  NO_VALID_DATA = 3;
}

message MessageCommon {
  bytes src_actor_id = 1;
  bytes dst_actor_id = 2;
  bytes queue_id = 3;
}

message StreamingQueueDataMsg {
  MessageCommon common = 1;
  uint64 seq_id = 2;
  uint64 msg_id_start = 3;
  uint64 msg_id_end = 4;
  uint64 length = 5;
  bool raw = 6;
}

message StreamingQueueCheckMsg {
  MessageCommon common = 1;
}

message StreamingQueueCheckRspMsg {
  MessageCommon common = 1;
  StreamingQueueError err_code = 2;
}

message StreamingQueueNotificationMsg {
  MessageCommon common = 1;
  uint64 seq_id = 2;
}

// for test
enum StreamingQueueTestRole {
  WRITER = 0;
  READER = 1;
}

message StreamingQueueTestInitMsg {
  StreamingQueueTestRole role = 1;
  bytes src_actor_id = 2;
  bytes dst_actor_id = 3;
  bytes actor_handle = 4;
  repeated bytes queue_ids = 5;
  repeated bytes rescale_queue_ids = 6;
  string test_suite_name = 7;
  string test_name = 8;
  uint64 param = 9;
}

message StreamingQueueTestCheckStatusRspMsg {
  string test_name = 1;
  bool status = 2;
}

message StreamingQueuePullRequestMsg {
  MessageCommon common = 1;
  uint64 msg_id = 2;
}

message StreamingQueuePullResponseMsg {
  MessageCommon common = 1;
  uint64 seq_id = 2;
  uint64 msg_id = 3;
  StreamingQueueError err_code = 4;
  bool is_upstream_first_pull = 5;
}

message StreamingQueueResendDataMsg {
  MessageCommon common = 1;
  uint64 first_seq_id = 2;
  uint64 last_seq_id = 3;
  uint64 seq_id = 4;
  uint64 msg_id_start = 5;
  uint64 msg_id_end = 6;
  uint64 length = 7;
  bool raw = 8;
}
